File: app/src/main/java/com/animsh/moviem/adapters/CastAdapter.java
package com.animsh.moviem.adapters;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.animsh.moviem.R;
import com.animsh.moviem.databinding.ItemCastContainerBinding;
import com.animsh.moviem.model.Cast;

import java.util.List;

/**
 * Created by animsh on 2/11/2021.
 */
public class CastAdapter extends RecyclerView.Adapter<CastAdapter.CastViewHolder> {

    private List<Cast> casts;
    private LayoutInflater layoutInflater;

    public CastAdapter(List<Cast> casts) {
        this.casts = casts;
    }

    @NonNull
    @Override
    public CastViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (layoutInflater == null) {
            layoutInflater = LayoutInflater.from(parent.getContext());
        }
        ItemCastContainerBinding castBinding = DataBindingUtil.inflate(
                layoutInflater, R.layout.item_cast_container, parent, false
        );
        return new CastViewHolder(castBinding);
    }

    @Override
    public void onBindViewHolder(@NonNull CastViewHolder holder, int position) {
        holder.bindCast(casts.get(position));
    }

    @Override
    public int getItemCount() {
        return casts.size();
    }

    class CastViewHolder extends RecyclerView.ViewHolder {

        private ItemCastContainerBinding itemCastContainerBinding;

        public CastViewHolder(ItemCastContainerBinding itemCastContainerBinding) {
            super(itemCastContainerBinding.getRoot());
            this.itemCastContainerBinding = itemCastContainerBinding;
        }

        public void bindCast(Cast cast) {
            itemCastContainerBinding.setCast(cast);
            itemCastContainerBinding.executePendingBindings();
        }
    }
}

File: app/src/main/java/com/animsh/moviem/adapters/CastAdapter.kt
package com.animsh.moviem.adapters

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.animsh.moviem.databinding.ItemCastContainerBinding
import com.animsh.moviem.models.movie.Cast
import com.animsh.moviem.util.MovieDiffUtil

/**
 * Created by animsh on 5/14/2021.
 */
class CastAdapter : RecyclerView.Adapter<CastAdapter.CrewViewHolder>() {

    private var castList = emptyList<Cast>()

    class CrewViewHolder(private val binding: ItemCastContainerBinding) :
        RecyclerView.ViewHolder(binding.root) {
        fun bindData(cast: Cast) {
            binding.cast = cast
            binding.executePendingBindings()
        }

        companion object {
            fun from(parent: ViewGroup): CrewViewHolder {
                val layoutInflater = LayoutInflater.from(parent.context)
                val binding = ItemCastContainerBinding.inflate(layoutInflater, parent, false)
                return CrewViewHolder(binding)
            }
        }
    }

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): CrewViewHolder {
        return CrewViewHolder.from(parent)
    }

    override fun onBindViewHolder(holder: CrewViewHolder, position: Int) {
        val current = castList[position]
        holder.bindData(current)
    }

    override fun getItemCount(): Int {
        return castList.size
    }

    fun setData(newData: List<Cast>) {
        val diffUtil = MovieDiffUtil(castList, newData)
        val diffUtilResult = DiffUtil.calculateDiff(diffUtil)
        castList = newData
        diffUtilResult.dispatchUpdatesTo(this)
    }

    fun clearList() {
        castList = emptyList()
    }
}